### PR TITLE
UseImportsTraverser: don't deep traverse

### LINKED ISSUE #8194: UseImportsTraverser optimization opportunity
looking at a codeigniter profile, I can see  `UseImportsTraverser` to use a considerable amount of time

grafik

In the `UseImportsTraverser` I can see it is using `SimpleCallableNodeTraverser` to deep travers statements.

looking at the [php-src docs](https://www.php.net/manual/en/language.namespaces.importing.php) I noted..

> The use keyword must be declared in the outermost scope of a file (the global scope) or inside namespace declarations. This is because the importing is done at compile time and not runtime, so it cannot be block scoped. The following example will show an illegal use of the use keyword

doesn't this mean, that we can stop traversing the statements in `UseImportsTraverser` as soon as we hit a function or class?
(or maybe it leaset it means we need only a linear scan of the top level and no deep traversal).

leaving it here for discussion

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/rectorphp/rector/issues/8194

before in codeigniter4

```
$ time php vendor/bin/rector process tests --dry-run --clear-cache
 326/326 [============================] 100%

 [OK] Rector is done!

real    1m3.584s
user    0m0.000s
sys     0m0.093s
```

after
```
time php vendor/bin/rector process tests --dry-run --clear-cache
 326/326 [============================] 100%

 [OK] Rector is done!

real    0m57.164s
user    0m0.015s
sys     0m0.031s
```

-> 10 % faster

### CODE REVIEW

File: rules/CodingStyle/ClassNameImport/UseImportsTraverser.php [hunk 1]
```diff
@@ -77,6 +79,10 @@ private function traverseForType(array $stmts, callable $callable, int $desiredT
                 $this->processGroupUse($node, $desiredType, $callable);
             }
 
+            if (!$node instanceof Namespace_) {
```
Comment (samsonasik) [inline rules/CodingStyle/ClassNameImport/UseImportsTraverser.php]: there are 2 top level node:

- `Rector\Core\PhpParser\Node\CustomNode\FileWithoutNamespace`
- `PhpParser\Node\Stmt\Namespace_`

File: rules/CodingStyle/ClassNameImport/UseImportsTraverser.php [hunk 2]
```diff
@@ -77,6 +80,10 @@ private function traverseForType(array $stmts, callable $callable, int $desiredT
                 $this->processGroupUse($node, $desiredType, $callable);
             }
 
+            if (!$node instanceof Namespace_ && !$node instanceof FileWithoutNamespace) {
```
Comment (samsonasik) [inline rules/CodingStyle/ClassNameImport/UseImportsTraverser.php]: How about just use `if else if else` ?

```php
if ($node instanceof Use_) {
   // ...
} elseif ($node instanceof GroupUse) {
  // ...
} else {
    return NodeTraverser::DONT_TRAVERSE_CHILDREN;
}

return null;
```

Comment (staabm) [reply rules/CodingStyle/ClassNameImport/UseImportsTraverser.php]: :+1:

**Review by samsonasik — APPROVED** [review]
Looking good, Trigger Rectify needs to be run on rector-src code base after this PR merged to ensure nothing oversight :)

**Comment by samsonasik** [comment]
Thank you @staabm

**Pull request merged**